scipy/linalg/_matfuncs_expm.c

```diff
@@ -0,0 +1,2375 @@
+#include "_matfuncs_expm.h"
+
+// For some reason these defs are not picked up from the header.
+#if defined(_MSC_VER)
+    // MSVC nonsense
+    #include <complex.h>
+    #define EXPM_Z _Dcomplex
+    #define EXPM_C _Fcomplex
+    #define CPLX_Z(real, imag) (_Cbuild(real, imag))
+    #define CPLX_C(real, imag) (_FCbuild(real, imag))
+#else
+    // C99-compliant compilers
+    #include <complex.h>
+    #define EXPM_Z double complex
+    #define EXPM_C float complex
+    #define CPLX_Z(real, imag) (real + imag*I)
+    #define CPLX_C(real, imag) (real + imag*I)
+#endif
+
+
+static float snorm1(float*, float*, const Py_ssize_t);
+static double dnorm1(double*, double*, const Py_ssize_t);
+static float cnorm1(EXPM_C*, float*, const Py_ssize_t);
+static double znorm1(EXPM_Z*, double*, const Py_ssize_t);
+static float snorm1est(float*, int);
+static double dnorm1est(double*, int);
+static float cnorm1est(EXPM_C*, int);
+static double znorm1est(EXPM_Z*, int);
+
+
+// SciPy's "Better than nothing", recursive matrix transpose for square arrays
```

Comment:
It also works for non-square arrays right?

Reply:
It might but then you need to keep track of the other dimension `m`. And it becomes a bit more hairy. My unicorn goal is that a language actually provides a fast transpose natively. There are omatcopy BLAS extensions but they are not supported in all BLAS libs